active/0016-emqx-conf.md

```diff
@@ -0,0 +1,184 @@
+
+#### emqx_conf application
+Before, the initialization of the configuration file, cluster_call, was done through the `emqx_machine`, we will split this part of the functionality from the` emqx_machine` and make a new application `emqx_conf`
+
+The role of this application is to
+- Converting the configuration from hocon format to erlang sys.config format at initialization.
+- Manage hot updates and deletions of configurations within the cluster.
+
+The basic node configuration hot update remains in emqx, in other words: `emqx_config_handler`, remains unchanged, but other apps that want to update the configuration must call through the `emqx_conf`'s' API, which cannot call emqx API directly.
+The specific flow is: 
+
+```bash
+Other Apps(eg: emqx_resource) => emqx_conf => emqx API.
+```
+
+### HTTP API design
+
+#### Get the whole configurations.
+
+```erlang
+#{
+  get => #{
+    description => <<"Get all the configurations of cluster, including hot and non-hot updatable items.">>,
+    parameters => [
+       {node, hoconsc:mk(typerefl:atom(),
+          #{in => query, required => false, example => <<"emqx@127.0.0.1">>,
+          desc => <<"Node's name: If you have this field defaulted, the whole cluster is return.">>})},
+       {debug, hoconsc:mk(typerefl:boolean(), #{in => query, required => false,
+          desc => <<"Carries debug(meta data) information, such as:line number,default value,documentation">>})}],
+            responses => #{
+                200 => #{"$node" => configs_list()}
+            }
+        }
+    };
+```
+
+- Returns what the current value/documentation of all configuration items is, group by nodename.
+-  `debug=true`, will return all the meta data, such as line number, default, document, easy to locate the problem.
+
+#### Update specific configuration
+
+```erlang
+schema("/configs/:rootname") ->
+    #{
+      get => #{
+            description => <<"Get the sub-configurations">>,
+            parameters => [
+              {node, hoconsc:mk(atom(), #{in => query, required => false})},
+              {debug, hoconsc:mk(typerefl:boolean(), #{in => query, required => false})}
+                          ],
+            responses => #{
+                200 => #{<<"$node">> => config_list()},
+                404 => emqx_dashboard_swagger:error_codes(['NOT_FOUND'], <<"config not found">>)
+            }
+        },
+        put => #{
+            description => <<"Update the sub-configurations">>,
+            parameters => [{node, hoconsc:mk(atom(), #{in => query, required => false})}],
+            requestBody => config_list(),
+            responses => #{
+                200 => #{<<"$node">> => config_list()},
+                400 => emqx_dashboard_swagger:error_codes(['UPDATE_FAILED'])
+            }
+        }
+    }.
+```
+
+- get specific configuation, such as: `/configs/emqx_dashboard` will return :
+
+  ```erlang
+  #{'emqx@127.0.0.1' => 
+     #{default_password => "public",
```

Comment:
There should be a `sensitive` flag in the schema for this field, and the value should be reported back as `"******"` in the API.